[FIX] web, product, project: allow reopening deletion confirmation dialog

yavy-odoo opened the pull request:
### Description of the issue/feature this PR addresses:

Behavior in <= v17.4:

The onclick method for the delete button in the one2many list calls the `onDeleteRecord` method. This is inherited by the model-specific list renderer components to insert the deletion confirmation dialog before calling super which sets the clicked data attribute to be true. In this case, we display the dialog before setting the button as clicked.

Behavior in v18:

The onclick method for the delete button in the one2many list calls the `onRemoveCellClicked`. This method sets the clicked as True and then calls the `onDeleteRecord` method that is inherited by the various model-specific list renderer components to insert the Confirmation Dialog. In these cases, the dialog is called after we set the 'clicked' data-attribute on the delete button to be True. Hence now we are unable to click on the same button again due to it being in the clicked state.

### Current behavior before PR:

[Deletion Dialog does not reopen](https://github.com/user-attachments/assets/29b6c2c7-553b-4745-a20b-26a470952a01)

1. Try to delete a record of Project Subtask or Product Attribute Value
2. This would open the Deletion Confirmation Dialog box
3. Click on cancel 
4. Try to delete the same record again 

Result: Nothing happens and the confirmation dialog does not show up again for the second time


### Desired behavior after PR is merged:

If user attempts to delete any record that opens the confirmation dialog and then cancels it, the user is able to delete it as the confirmation dialog box is no longer disabled after spawning once.


opw-4731375

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/207621.png)](https://mergebot.odoo.com/odoo/odoo/pull/207621)

Guillaume-gdi reviewed (changes requested):
This change is too risky for a stable version IMO. This could break custom code. If there is no way to make this fix stable compliant, I would retarget the PR on the master version.

Could be nice to add a test too.

yavy-odoo commented:
Hi @Guillaume-gdi 

Thank you for your review! I have updated this PR with the following changes:
1. Changed target branch to odoo:master
2. Added a Hoot (JS) test to verify that we can re-open the deletion dialog for Project Sub-task list view after clicking on cancel

Could you please re-review this? Thanks

Guillaume-gdi reviewed (commented):
LGTM :+1: Thank's for adding the test.

aab-odoo commented:
@yavy-odoo I don't think that's the way to go. You basically bypass the faulty code in web. I pinged @abz-odoo on this, he'll take a look and try to make our previous fix more robust, s.t. it doesn't break in the product and project usecases.

aab-odoo commented on addons/project/static/tests/project_task_subtask.test.js:
```diff
@@ -313,3 +313,24 @@ test("project.task (kanban): check subtask creation when input is empty", async
+    expect(`.modal-footer .btn-primary:contains(Ok)`).toHaveCount(1, {
+        message: "Delete action should be visible on reopening the dialog",
+    });
+});
```
Ok then

yavy-odoo commented:
Hi @aab-odoo 

Apologies for the delay but I have implemented the following feedback to this PR and wanted to request you to re-review this once you have time. Here's a summary of what I fixed:

1. The PR now targets v18
2. We are not removing the onRemoveCellClicked method since it may break customization on v18 and we would later change the PR targetting master to remove it correctly.

Here's a short demo of me testing the updated code:
https://github.com/user-attachments/assets/f40df038-4ed5-41c9-850f-70261bf8a60d


Thank you!

aab-odoo reviewed (commented):
you can squash both commit I think.

Also, runbot is red

aab-odoo commented on addons/web/static/src/views/list/list_renderer.js:
```diff
@@ -96,6 +97,11 @@ export class ListRenderer extends Component {
         this.tableRef = useRef("table");
+        this.debouncedOnDeleteRecord = debounce(
+            this.onDeleteRecord.bind(this),
+            50,
```
300 was fine

yavy-odoo replied:
Thank for the feedback!
The reason behind changing this was purely debugging the failing tests. I noticed that previously we had a lot more failing tests:
1. Before, using 300ms: https://runbot.odoo.com/runbot/build/84147707 which most likely happened because the debounce timeout was too long for the test steps that covered clicking the delete button
2. After changing it to 50ms, the number of failing test suites are way lesser: https://runbot.odoo.com/runbot/build/84950736 (I am yet to fix the ones that are still failing here)

Another point that I observed was the original commit that attempted to previously fix this bug added a test case: https://github.com/odoo-dev/odoo/commit/a02e497c4b3ef2bfbf4f8396a0dbfc0cc90b59ae#diff-994b09b85a8f1df25e39a78389d038215dbfea7e07ed2a607b2b7147f171927f to handle the issue happening due to the double click and this test is not among the failing ones.

I fear that having a high debounce delay is the reason behind a lot of existing tests failing

aab-odoo replied:
This shows that this PR is likely to introduce non determinism in tests, so we must be careful, and probably ask the runbot team for multi builds at some point, to be sure that it doesn't.

I would increase the delay (even more than 300), to see all tests that fail. Then, read each one of them and understand why it fails. I doubt that any test, except the one introduced by abz [here](https://github.com/odoo-dev/odoo/commit/a02e497c4b3ef2bfbf4f8396a0dbfc0cc90b59ae#diff-994b09b85a8f1df25e39a78389d038215dbfea7e07ed2a607b2b7147f171927f), purposely double clicks on the same trash icon. A common reason for failing tests might be the successive removal of 2 different records. The second click wouldn't work if it comes too soon after the first one. We could handle this usecase by debouncing callbacks for each record individually if necessary. Otherwise, we could as well wait for the delay in all impacted tests.
